File: 6_print_link_reversely/go/main.go
package main

import (
	"fmt"
)

type listNode struct {
	val  int
	next *listNode
}

func printList(head *listNode) {
	nodes := make([]*listNode, 0)
	p := head
	for p != nil {
		nodes = append(nodes, p)
		p = p.next
	}
	for i := len(nodes) - 1; i >= 0; i-- {
		fmt.Printf("%d ", nodes[i].val)
	}
	fmt.Printf("\n")
}

func printList2(head *listNode) {
	if head == nil {
		return
	}
	if head.next != nil {
		printList2(head.next)
	}
	fmt.Printf("%d ", head.val)
}

func main() {
	node1 := &listNode{1, nil}
	node2 := &listNode{2, nil}
	node1.next = node2
	node3 := &listNode{3, nil}
	node2.next = node3

	printList(node1)
	printList2(node1)
	fmt.Printf("\n")
}

File: 6_print_link_reversely/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, val):
        self.val = val
        self.next = None


def print_list(head):
    nodes = []
    p = head
    while p is not None:
        nodes.append(p)
        p = p.next
    for p in nodes[::-1]:
        print(p.val),
    print('')


def print_list_recursively(head):
    if head is not None:
        if head.next is not None:
            print_list_recursively(head.next)
        print(head.val),


if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    node3 = ListNode(3)
    node2.next = node3

    print_list(node1)
    print_list_recursively(node1)
